File: backend/app/schemas/documents.py
from datetime import datetime
from typing import Annotated, Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

Text = Annotated[
    str,
    StringConstraints(strip_whitespace=True, strict=True, min_length=1, max_length=254),
]
LongText = Annotated[
    str,
    StringConstraints(strip_whitespace=True, strict=True, min_length=1, max_length=5_000),
]
SearchQuery = Annotated[
    str,
    StringConstraints(strip_whitespace=True, strict=True, min_length=1, max_length=100),
]
OptionCode = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        strict=True,
        min_length=1,
        max_length=64,
        pattern=r"^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$",
    ),
]

# public.file 의 file_processing_status_check 와 같은 목록이어야 한다.
# 한 값이라도 빠지면 그 값이 든 행을 담은 목록 응답 전체가 검증에서 터진다.
ProcessingStatus = Literal["uploaded", "processing", "review_required", "completed", "failed"]


class _WriteModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class DocumentCreate(_WriteModel):
    category_code: OptionCode
    title: Text
    # 화면에서는 '메모' 다. 자료가 무엇인지 적는 한 칸이라 컬럼은 그대로 쓴다.
    description: LongText | None = None
    # 고객사와 발주는 새로 고를 수 없지만, 예전 자료가 들고 있어 쓰기는 열어 둔다.
    customer_company_id: UUID | None = None
    customer_contact_id: UUID | None = None
    sales_deal_id: UUID | None = None
    purchase_order_id: UUID | None = None
    product_id: UUID | None = None


class DocumentPatch(_WriteModel):
    category_code: OptionCode | None = None
    title: Text | None = None
    description: LongText | None = None
    customer_company_id: UUID | None = None
    customer_contact_id: UUID | None = None
    sales_deal_id: UUID | None = None
    purchase_order_id: UUID | None = None
    product_id: UUID | None = None


class DocumentFileRead(BaseModel):
    """storage_key 는 내부 저장소 주소라 내보내지 않는다."""

    id: UUID
    version_no: int
    file_name: str
    media_type: str | None
    byte_size: int
    processing_status: ProcessingStatus
    processing_error: str | None = None
    uploaded_by_member_id: UUID
    uploaded_by_display_name: str
    note: str | None
    uploaded_at: datetime


class DocumentProcessBatchRequest(_WriteModel):
    """자료실에서 서버 처리를 시작할 파일 목록."""

    file_ids: list[UUID] = Field(min_length=1, max_length=50)


class DocumentProcessBatchRead(BaseModel):
    """서버가 처리를 접수한 파일별 현재 상태."""

    files: list[DocumentFileRead]


class DocumentSummaryRead(BaseModel):
    file_id: UUID
    file_name: str
    processing_status: ProcessingStatus
    processing_error: str | None
    extracted_text: str | None
    extracted_markdown: str | None
    extracted_payload: dict[str, Any] | None
    summary_markdown: str | None
    summary_payload: dict[str, Any] | None
    processed_at: datetime | None


class DocumentChunkRead(BaseModel):
    chunk_id: UUID
    document_id: UUID
    file_id: UUID
    chunk_no: int
    page_start: int | None
    page_end: int | None
    section: str | None
    content: str
    score: float
    metadata: dict[str, Any]


class DocumentBriefingSourceRead(BaseModel):
    """영업·계약관리 Agent가 브리핑 근거로 사용할 RAG 검색 결과."""

    chunk_id: UUID
    document_id: UUID
    file_id: UUID
    file_name: str
    chunk_no: int
    page_start: int | None = None
    page_end: int | None = None
    section: str | None
    content: str
    score: float
    metadata: dict[str, Any]


class DocumentBriefingSummaryRead(BaseModel):
    """검색된 자료에 저장된 구조화 요약."""

    file_id: UUID
    document_id: UUID
    file_name: str
    summary_markdown: str
    summary_payload: dict[str, Any] | None


class DocumentBriefingContextRead(BaseModel):
    """브리핑 생성 Agent가 한 번에 소비할 자료요약·RAG 묶음."""

    query: str
    summaries: list[DocumentBriefingSummaryRead]
    sources: list[DocumentBriefingSourceRead]


class DocumentRead(BaseModel):
    id: UUID
    document_no: str
    category_code: str
    title: str
    description: str | None
    customer_company_id: UUID | None
    customer_company_name: str | None
    customer_contact_id: UUID | None
    sales_deal_id: UUID | None
    # 목록의 연결 칸이 id 대신 사람이 읽을 값을 보이게 이름을 함께 준다.
    sales_deal_no: str | None
    purchase_order_id: UUID | None
    product_id: UUID | None
    product_name: str | None
    created_by_member_id: UUID
    created_by_display_name: str
    created_at: datetime
    files: list[DocumentFileRead]
    latest_version_no: int | None


class DocumentUploaderRead(BaseModel):
    member_id: UUID
    display_name: str


class DocumentPage(BaseModel):
    items: list[DocumentRead]
    skip: int
    limit: int
    total: int
    has_more: bool
    next_skip: int | None
    # 분류 탭 옆 건수 {분류 코드: 건수}. 고른 분류는 빼고 센 값이라 total 과 다르다.
    counts: dict[str, int] = Field(default_factory=dict)
    # 담당자 고르는 칸에 세울 사람. 최신 버전을 올린 사람 기준이고, 쪽에 담긴 문서만
    # 보면 지금 쪽에 없는 사람을 고를 수 없어서 서버가 전체에서 뽑아 준다.
    uploaders: list[DocumentUploaderRead] = Field(default_factory=list)


class DocumentPageParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    q: SearchQuery | None = None
    category_code: list[OptionCode] | None = None
    customer_company_id: UUID | None = None
    sales_deal_id: UUID | None = None
    created_by_member_id: list[UUID] | None = None
    # 아래 둘은 최신 버전 파일을 본다. 화면의 담당자·기간 필터가 "마지막으로 올린 사람과
    # 올린 날짜" 를 뜻하기 때문이다. 문서를 처음 만든 사람(created_by_member_id)과 다르다.
    latest_uploader_member_id: list[UUID] | None = None
    latest_uploaded_from: datetime | None = None
    skip: int = Field(default=0, ge=0, le=9_223_372_036_854_775_807)
    limit: int = Field(default=30, ge=1, le=30)


class DownloadRead(BaseModel):
    """짧게 사는 다운로드 주소. 매 요청마다 권한을 다시 검사하고 발급한다."""

    url: str
    expires_in: int
    file_name: str
    media_type: str | None
    byte_size: int

File: frontend/src/types/documents.ts
/** 자료실 문서의 분류. 탭과 정렬 순서는 Documents/catalog.ts 가 정합니다. */
export type DocumentCategory = '계약서' | '발주서' | '상품설명서' | '견적서' | '기타'

/** 파일 종류. 확장자를 이만큼으로 뭉쳐 배지 하나로 보여 줍니다. */
export type DocumentFileKind = 'pdf' | 'doc' | 'sheet' | 'slide' | 'image' | 'etc'

/**
 * 문서가 붙어 있는 대상. 아무 데도 붙지 않은 문서는 kind 가 'none' 입니다.
 *
 * 새로 고를 수 있는 것은 '상품' 과 '딜' 뿐입니다. '고객사' 와 '발주' 는 예전에 붙인
 * 자료가 목록에서 연결을 잃지 않도록 읽기 쪽에만 남겨 둡니다.
 */
export interface DocumentLink {
  kind: 'none' | '상품' | '딜' | '고객사' | '발주'
  /** 연결 대상의 id. 고르는 순간 손에 들어옵니다. kind 가 'none' 이면 빈 문자열입니다. */
  id: string
  /** '초음파 진단기' 또는 'SL-DL-2026-0038'. kind 가 'none' 이면 빈 문자열입니다. */
  label: string
}

/** 자료 하나에 딸린 파일 하나. */
export interface DocumentFile {
  /** 아직 올리지 않은 자료의 자리끼움에는 없습니다. */
  id?: string
  documentId?: string
  fileName: string
  /** 바이트. 화면에 찍을 때 sizeLabel() 로 바꿉니다. */
  bytes: number
  owner: string
  /** YYYY-MM-DD */
  uploaded: string
  /** 올릴 때 적은 한 줄 */
  note: string
  processingStatus?: DocumentProcessingStatus
  processingError?: string | null
}

/** 자료실 문서 */
export interface SalesDocument {
  id: string
  documentNo?: string
  title: string
  category: DocumentCategory
  kind: DocumentFileKind
  link: DocumentLink
  /** 화면의 '메모'. 목록에서 이 자료가 무엇인지 알아볼 한 줄입니다. */
  description: string
  file: DocumentFile
}

export type DocumentProcessingStatus =
  'uploaded' | 'processing' | 'review_required' | 'completed' | 'failed'

export interface DocumentFileResponse {
  id: string
  file_name: string
  media_type: string | null
  byte_size: number
  processing_status: DocumentProcessingStatus
  processing_error?: string | null
  uploaded_by_member_id: string
  uploaded_by_display_name: string
  note: string | null
  uploaded_at: string
}

export interface DocumentSummaryResponse {
  file_id: string
  file_name: string
  processing_status: DocumentProcessingStatus
  processing_error: string | null
  extracted_text: string | null
  extracted_markdown: string | null
  extracted_payload: Record<string, unknown> | null
  summary_markdown: string | null
  summary_payload: Record<string, unknown> | null
  processed_at: string | null
}

export interface DocumentResponse {
  id: string
  document_no: string
  category_code: string
  title: string
  description: string | null
  customer_company_id: string | null
  customer_company_name: string | null
  sales_deal_id: string | null
  sales_deal_no: string | null
  purchase_order_id: string | null
  product_id: string | null
  product_name: string | null
  created_by_member_id: string
  created_by_display_name: string
  created_at: string
  file: DocumentFileResponse | null
}

export interface DownloadResponse {
  url: string
  expires_in: number
  file_name: string
  media_type: string | null
  byte_size: number
}
